utils/zerovec/src/ule/niche.rs

```diff
@@ -0,0 +1,184 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::mem::size_of;
+use core::ops::DerefMut;
+use core::{marker::Copy, ops::Deref};
+
+use super::{AsULE, ULE};
+
+/// The types implementing this trait guarantee that [`NicheBytes::INVALID_BIT_PATTERN`]
+/// can never occur as a valid byte representation of the type.
+/// The implementors guarantee that N == core::mem::sizeo_of::<Self>()
+pub trait NicheBytes<const N: usize> {
+    const INVALID_BIT_PATTERN: [u8; N];
```

Comment:
Discussion: Can we come up with a better name? I don't think "invalid" is exactly the right term in this context.

```suggestion
    const NICHE_BIT_PATTERN: [u8; N];
```

Reply:
Renamed to `NICHE_BIT_PATTERN`